src/gui/transferlistmodel.cpp

```diff
@@ -295,29 +294,51 @@ BitTorrent::TorrentHandle *TransferListModel::torrentHandle(const QModelIndex &i
 {
     if (!index.isValid()) return nullptr;
 
-    return m_torrents.value(index.row());
+    return m_torrentList.value(index.row());
 }
 
 void TransferListModel::handleTorrentAboutToBeRemoved(BitTorrent::TorrentHandle *const torrent)
 {
-    const int row = m_torrents.indexOf(torrent);
-    if (row >= 0) {
-        beginRemoveRows(QModelIndex(), row, row);
-        m_torrents.removeAt(row);
-        endRemoveRows();
+    const int row = m_torrentMap.value(torrent, -1);
+
+    if (row < 0)
+        return;
+
+    beginRemoveRows({}, row, row);
+
+    m_torrentList.removeAt(row);
+
+    m_torrentMap.remove(torrent);
+    for (int &value : m_torrentMap) {
+        if (value > row)
+            --value;
     }
+
+    endRemoveRows();
 }
 
 void TransferListModel::handleTorrentStatusUpdated(BitTorrent::TorrentHandle *const torrent)
 {
-    const int row = m_torrents.indexOf(torrent);
-    if (row >= 0)
-        emit dataChanged(index(row, 0), index(row, columnCount() - 1));
+    const int row = m_torrentMap.value(torrent, -1);
+
+    if (row < 0)
+        return;
+
+    emit dataChanged(index(row, 0), index(row, columnCount() - 1));
 }
 
-void TransferListModel::handleTorrentsUpdated()
+void TransferListModel::handleTorrentsUpdated(const QVector<BitTorrent::TorrentHandle *> &torrents)
 {
-    emit dataChanged(index(0, 0), index(rowCount() - 1, columnCount() - 1));
+    const int columns = (columnCount() - 1);
+
+    for (BitTorrent::TorrentHandle *const torrent : torrents) {
+        const int row = m_torrentMap.value(torrent, -1);
+
+        if (row < 0)
+            continue;
+
+        emit dataChanged(index(row, 0), index(row, columns));
```

Comment:
Are you sure N dataChanged will be faster than single dataChanged? Did you measure it?

Reply:
No I haven't measured it.
Doesn't dataChanged() causes the model to load the updated data? And specifying all data has changed and needs to reload is obviously slower than only updating the changed ones.

Reply:
>And specifying all data has changed and needs to reload is obviously slower than only updating the changed ones IMO.

Unfortunately, it may not be as obvious as it seems at first glance.There is some overhead from many signals emitting and so on. IMO, it needs to measure performance.

Reply:
>IMO, it needs to measure performance.

Well from my rudimentary test, 8 torrents with 4 downloading, 4 paused. The old code uses 1194980 nanoseconds and the new one here uses 155720 ns. I even bypassed the lookup and the results are still similar.
Seems I was right about the "more data to reload" theory.

Reply:
>I even bypassed the lookup and the results are still similar.

Your lookup optimization not affects with so small torrents number.

>The old code (reload all) uses 1194980 nanoseconds on avg. and the new one here uses 155720 ns on avg. 

Strange result... Theoretically, it should have been about twice as different (old code updates all 8 torrents, new code updates 4 unpaused torrents), but here the difference is an order of magnitude.
I need to test it myself.

Reply:
Well, I finished my tests.
I was mainly interested in corner cases, or rather, one corner case, when all torrents are updated.
As I expected, the new code has an overhead compared to the old one (when total number of torrents is more than two). And this overhead increases with the total number of torrents (it's about 30% for 60 torrents and it's more than 100% for 400 torrents).
So we need to have a solid understanding of how it behaves in the real world. What is the probability that a large number of torrents will be updated at once (relative to their total number)?
There may be (theoretically) a problem with those users who collect a lot of torrents (several thousand). Of course, there is no problem if the vast majority of them are just "dusting on the shelf" suspended. But I find it difficult to estimate the use case of such users to be sure that it is.

Reply:
Then I suppose there can be a threshold to switch between update all or the ones in QVector.
Something like:
```c++
const qreal threshold = 0.8;  // exact value to be decided
if (torrents.size() > (m_torrentList * threshold))  // exact formula to be decided
    emit dataChanged(index(0, 0), index(rowCount() - 1, columnCount() - 1));
else
    // emit dataChanged() for each one in `torrents`
```
I'll need your opinion on the formula & the threshold value.

Reply:
>I'll need your opinion on the formula & the threshold value.

I don't think we can anticipate a possible overhead in any static way. It is generated somewhere in the depths of the Qt code and can vary on different systems/Qt versions etc.

Reply:
We can either break your brain trying to calculate and control it in runtime, or just proceed from the idea that generally the number of active torrents is much less than passive ones and use the new way (we can also add the ability to switch to the old way somewhere in advanced settings).

Reply:
>or just proceed from the idea that generally the number of active torrents is much less than passive ones and use the new way 

That is the idea from my code snippet... I've adopted a naive way to do it.

>we can also add the ability to switch to the old way somewhere in advanced settings

I would consider it to be another PR.

Reply:
>I've adopted a naive way to do it.

Ok.
>I would consider it to be another PR.

Let's wait for "torrent collectors" feedback. Maybe you won't need it at all.

Reply:
```c++
int startRow = -1;
int endRow = -1;
for (int row = 0; row < torrentList.count(); ++row) {
    BitTorrent::TorrentHandle *const torrent = torrentList[row];
    if (torrents.contains(torrent)) {
        if (startRow == -1)
            startRow = row;
        endRow = row;
    }
    else {
        if (startRow >= 0) {
            emit dataChanged(index(startRow, 0), index(endRow, columns));
            startRow = -1;
            endRow = -1;
        }
    }
}
```
Algorithm that emits only one signal for each continuous range of rows. But 
`torrents` should be QSet for better lookup performance.

Reply:
I've thought about merging the row numbers, but I think it performs weak in the worst case and decided not worthy of the complexity it brings.

Reply:
Ok, let's start without it. Maybe there's a better algorithm later.